commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/hull/ConvexHull3D.java

```diff
@@ -0,0 +1,685 @@
+
+            // The point set is degenerate because all points are colinear.
+            if (line.contains(vertex3)) {
+                return new Simplex(Collections.emptyList());
+            }
+
+            // Form a triangle with the first three vertices.
+            ConvexPolygon3D facet1 = Planes.triangleFromVertices(vertex1, vertex2, vertex3, precision);
+
+            // Find a point with maximal distance to the plane formed by the triangle.
+            Plane plane = facet1.getPlane();
+            Vector3D vertex4 = points.stream()
+                    .max((u, v) -> Double.compare(Math.abs(plane.offset(u)), Math.abs(plane.offset(v)))).get();
+
+            // The point set is degenerate, because all points are coplanar.
+            if (plane.contains(vertex4)) {
+                return new Simplex(Collections.emptyList());
+            }
+
+            // Construct the other three facets.
+            ConvexPolygon3D facet2 = Planes.convexPolygonFromVertices(Arrays.asList(vertex1, vertex2, vertex4),
+                    precision);
+            ConvexPolygon3D facet3 = Planes.convexPolygonFromVertices(Arrays.asList(vertex1, vertex3, vertex4),
+                    precision);
+            ConvexPolygon3D facet4 = Planes.convexPolygonFromVertices(Arrays.asList(vertex2, vertex3, vertex4),
+                    precision);
+
+            List<Facet> facets = new ArrayList<>();
+
+            // Choose the right orientation for all facets.
+            facets.add(isInside(facet1, vertex4, precision) ? new Facet(facet1, precision) :
+                new Facet(facet1.reverse(), precision));
+            facets.add(isInside(facet2, vertex3, precision) ? new Facet(facet2, precision) :
+                new Facet(facet2.reverse(), precision));
+            facets.add(isInside(facet3, vertex2, precision) ? new Facet(facet3, precision) :
+                new Facet(facet3.reverse(), precision));
+            facets.add(isInside(facet4, vertex1, precision) ? new Facet(facet4, precision) :
+                new Facet(facet4.reverse(), precision));
+
+            return new Simplex(facets);
+        }
+
+        /**
+         * Returns {@code true} if the given point resides inside the negative
+         * half-space of the oriented facet. Points which are coplanar are also assumed
+         * to be inside. Mathematically a point is inside if the calculated oriented
+         * offset, of the point to the hyperplane is less than or equal to zero.
+         *
+         * @param facet     the given facet.
+         * @param point     a reference point.
+         * @param precision the given precision.
+         * @return {@code true} if the given point resides inside the negative
+         *         half-space.
+         */
+        private static boolean isInside(ConvexPolygon3D facet, Vector3D point, DoubleEquivalence precision) {
```

Comment:
Yes, the only reason it isn't right now is because of line 345f. where it is used before creating the facet. I could move this in the Facet class, but this would mean that we have to create a facet first and then determine if we need to reverse the orientation, which would potentially mean that we have to create two objects.

Reply:
I changed the constructor of Facet to handle this logic and moved the method there. Since it was then always negated I also inverted the calculation and renamed the method to `isOutside`

Reply:
The method is no longer present. Due to other changes it is no an offset method in the Facet class.